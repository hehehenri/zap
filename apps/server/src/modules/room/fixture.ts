import { RoomDefinition, RoomModel } from "./RoomModel";

type Fields = Pick<RoomDefinition, Exclude<keyof RoomDefinition, "_id">>

export const createRoom = async (fields?: DeepPartial<Fields>) => {
  return await new RoomModel(fields).save();
}
