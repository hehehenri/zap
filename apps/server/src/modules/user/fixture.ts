import config from "../../config";
import { UserDefinition, UserModel } from "./UserModel"
import bcrypt from "bcryptjs";

let counter = 0;

const getCounter = () => {
  counter += 1;

  return counter;
}

type Fields = Pick<UserDefinition, Exclude<keyof UserDefinition, "_id">>

export const createUser = async (fields?: DeepPartial<Fields>) => {
  const index = getCounter();

  const plainTextPassword = fields?.password ?? "password";

  const password = bcrypt.hashSync(plainTextPassword, config.jwt.saltRounds);
  const username = fields?.username ?? `user#${index}`;

  const user = new UserModel({ password, username });
  await user.save();

  return user;
}
