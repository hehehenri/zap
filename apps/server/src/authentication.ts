import jwt from "jsonwebtoken";
import { UserModel, UserDocument } from "./modules/user/UserModel";
import config from "./config";
import { KoaContext } from "./context";

export type Config = {
  secret: string
}

type TokenPayload = {
  id: string
}

export const getToken = (ctx: KoaContext) => {
  const auth = ctx.headers.authorization;
  if (!auth) return null;

  if (auth.startsWith("Bearer ")) {
    return auth.substring(7, auth.length);
  }

  return null
}

export const getAuth = async (token: string | null) => {
  if (!token) return { user: null };

  try {
    const decoded = jwt.verify(token, config.jwt.secret) as TokenPayload;

    const user = await UserModel.findById(decoded.id);
    return { user };
  } catch (_) {
    return { user: null };
  }
}

export const generateToken = (user: UserDocument) => {
  return jwt.sign({ id: user._id }, config.jwt.secret);
}
