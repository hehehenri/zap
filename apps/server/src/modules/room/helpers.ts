import { UserDocument } from "../user/UserModel";
import { RoomDocument } from "./RoomModel";

export const isRoomMember = (room: RoomDocument, user: UserDocument) => {
  return room
    .participants
    .map(u => u._id.toString())
    .includes(user._id.toString());
}
