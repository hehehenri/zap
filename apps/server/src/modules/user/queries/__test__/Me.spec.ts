import { generateToken } from "../../../../authentication";
import { createUser } from "../../fixture";
import { describeWithDb, testQuery } from "../../../../test/helpers";

const query = {
  query: `
    query Me {
      me { username }
    }
  `,
  variables: {}
};

describeWithDb('user/queries/me', () => {  
  it("should return logged user", async () => {
    const user = await createUser();
    const token = generateToken(user);

    const response = await testQuery({ query, token });

    expect(response.status).toBe(200);
    expect(response.body.data).toMatchSnapshot();
  });

  it(
    "should return unauthorized error when token is not provided",
    async () => {
      const response = await testQuery({ query });

      expect(response.status).toBe(401);
      expect(response.body.errors).toMatchSnapshot();
    }
  );
});
