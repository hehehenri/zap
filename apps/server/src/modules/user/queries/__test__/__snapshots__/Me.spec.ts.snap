// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user/queries/me should return logged user 1`] = `
{
  "me": {
    "username": "user#1",
  },
}
`;

exports[`user/queries/me should return unauthorized error when token is not provided 1`] = `
[
  {
    "locations": [
      {
        "column": 7,
        "line": 3,
      },
    ],
    "message": "User not authorized",
    "path": [
      "me",
    ],
  },
]
`;
